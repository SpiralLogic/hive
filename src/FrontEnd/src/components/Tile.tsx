import { FunctionComponent, h } from 'preact';
import { Tile } from '../domain';
import { TileAction } from '../utilities/hive-dispatcher';
import { addHiveEventListener, useClassReducer, useHiveDispatcher } from '../utilities/hooks';
import { deepEqual } from 'fast-equals';
import { handleDrop, handleKeyboardNav, isEnterOrSpace } from '../utilities/handlers';
import { memo } from 'preact/compat';
import { useEffect, useState } from 'preact/hooks';

const tileSelector = `[tabindex].tile`;
const cellSelector = `[tabindex].cell`;
const playerSelector = `[tabindex].name`;

const TileFC: FunctionComponent<Tile> = (tile: Tile) => {
  const { id, moves, creature, playerId } = tile;
  const [focus, setFocus] = useState(tileSelector);
  const [classes, setClassList] = useClassReducer(`player${playerId} hex tile`);
  const hiveDispatcher = useHiveDispatcher();

  const deselect = () => {
    if (!classes.includes('selected')) return;
    setClassList({ type: 'remove', class: 'selected' });
    hiveDispatcher.dispatch({ type: 'tileDeselected', tile: tile });
  };

  const select = () => {
    if (classes.includes('selected')) return;
    hiveDispatcher.dispatch({ type: 'tileClear' });
    setClassList({ type: 'add', class: 'selected' });
    hiveDispatcher.dispatch({ type: 'tileSelected', tile: tile });
  };

  addHiveEventListener<TileAction>('tileSelect', (event: TileAction) => {
    if (event.tile.id == id) select();
  });

  addHiveEventListener<TileAction>('tileDeselect', (event: TileAction) => {
    if (event.tile.id == id) deselect();
  });

  addHiveEventListener('tileClear', () => {
    deselect();
    setFocus('');
  });

  const handleDragStart = (event: DragEvent) => {
    event.stopPropagation();
    select();
    setClassList({ type: 'add', class: 'beforeDrag' });
    setTimeout(() => setClassList({ type: 'remove', class: 'beforeDrag' }), 1);
  };
  const handleDragEnd = () => {
    hiveDispatcher.dispatch({ type: 'tileClear' });
    hiveDispatcher.dispatch({ type: 'tileDropped', tile: tile });
  };
  const handleClick = (event: MouseEvent) => {
    event.stopPropagation();
    !classes.includes('selected') ? select() : deselect();
  };
  const handleKeyDown = (e: KeyboardEvent) => {
    if (handleKeyboardNav(e) || !isEnterOrSpace(e)) return;
    e.stopPropagation();
    if (!classes.includes('selected')) {
      select();
      setFocus(cellSelector);
    } else {
      deselect();
    }
  };
  const handleMouseLeave = (event: { currentTarget: HTMLElement }) => event.currentTarget.blur();

  useEffect(() => {
    if (!focus) return;
    const focusElement =
      document.querySelector<HTMLElement>(focus) || document.querySelector<HTMLElement>(playerSelector);
    focusElement?.focus();
    setFocus('');
  }, [focus]);

  const attributes = {
    title: creature,
    class: classes,
    draggable: !!moves.length,
    tabindex: moves.length ? 0 : undefined,
  };
  const handlers = moves.length
    ? {
        onclick: handleClick,
        ondragstart: handleDragStart,
        ondragend: handleDragEnd,
        onkeydown: handleKeyDown,
        onmouseleave: handleMouseLeave,
        ondrop: handleDrop,
      }
    : {
        ondrop: handleDrop,
      };
  return (
    <div {...attributes} {...handlers}>
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 100 100">
        <use href="#hex" />
        <use class="creature" href={`#${creature.toLowerCase()}`} />
      </svg>
    </div>
  );
};
TileFC.displayName = 'Tile';
export default memo(TileFC, deepEqual);
