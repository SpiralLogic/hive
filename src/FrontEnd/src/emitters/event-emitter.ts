export class EventEmitter<TEvent> {
  private listeners = new Set<EventListener<TEvent>>();

  add = (...listeners: EventListener<TEvent>[]): void => {
    listeners.forEach((l) => this.listeners.add(l));
  };

  emit = (event: TEvent) => {
    this.listeners.forEach((l): void => l(event));
  };

  remove = (...listeners: EventListener<TEvent>[]): void => {
    listeners.forEach((l) => this.listeners.delete(l));
  };
}

type EventListener<TEvent> = (event: TEvent) => void;
