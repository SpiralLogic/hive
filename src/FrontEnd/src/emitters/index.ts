export * from './cell-event-emitter';
export * from './tile-event-emitter';
