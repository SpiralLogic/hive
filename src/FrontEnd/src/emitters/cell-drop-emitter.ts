import { EventEmitter } from './event-emitter';
import { Move } from '../domain';

export class CellDropEmitter extends EventEmitter<CellDropEvent> {}

export type CellDropListener<TCellDropEvent> = (event: TCellDropEvent) => void;

const cellDropEmitter = new CellDropEmitter();

export type CellDropEvent = { type: 'drop'; move: Move };

export const useCellDropEmitter = (): CellDropEmitter => cellDropEmitter;
