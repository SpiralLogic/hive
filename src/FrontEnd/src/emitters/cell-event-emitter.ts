import { EventEmitter } from './event-emitter';
import { Move } from '../domain';

export class CellEventEmitter extends EventEmitter<CellEvent> {}

export type CellEventListener<TEvent> = (event: TEvent) => void;

const cellEventEmitter = new CellEventEmitter();

export type CellEvent = { type: 'drop' | 'deselect' };
export type CellMoveEvent = CellEvent & { move: Move };

export const useCellEventEmitter = (): CellEventEmitter => cellEventEmitter;
