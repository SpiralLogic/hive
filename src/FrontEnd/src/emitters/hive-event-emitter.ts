import { EventEmitter } from './event-emitter';
import { Inputs, useEffect } from 'preact/hooks';
import { Move, Tile } from '../domain';

export class HiveEventEmitter extends EventEmitter<HiveEvent> {}

export type HiveEventListener<TEvent> = (event: TEvent) => void;

const hiveEventEmitter = new HiveEventEmitter();

export type HiveEvent = MoveEvent | TileEvent | { type: 'deselect' };
export type MoveEvent = { type: 'move'; move: Move };
export type TileEvent = { type: 'start' | 'end' | 'click'; tile: Tile };

export const useHiveEventEmitter = (
  hiveEventHandler?: HiveEventListener<HiveEvent>,
  inputs?: Inputs
): HiveEventEmitter => {
  if (hiveEventHandler)
    useEffect(() => {
      hiveEventEmitter.add(hiveEventHandler);
      return () => {
        hiveEventEmitter.remove(hiveEventHandler);
      };
    }, inputs);

  return hiveEventEmitter;
};
