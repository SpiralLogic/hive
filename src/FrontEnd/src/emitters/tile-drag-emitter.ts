import {EventEmitter} from './event-emitter';
import {Tile} from '../domain';

export type TileDragEvent = {
    type: 'start' | 'end' | 'click';
    tile: Tile;
};

export class TileDragEmitter extends EventEmitter<TileDragEvent> {
}

const tileDragEmitter = new TileDragEmitter();

export const useTileDragEmitter = () => tileDragEmitter;
