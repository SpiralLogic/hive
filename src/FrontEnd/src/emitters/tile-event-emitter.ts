import {EventEmitter} from './event-emitter';
import {Tile} from '../domain';

export type TileEvents = {
    type: 'start' | 'end' | 'click';
    tile: Tile;
};

export class TileEventEmitter extends EventEmitter<TileEvents> {
}

const tileEventEmitter = new TileEventEmitter();

export const useTileEventEmitter = () => tileEventEmitter;
