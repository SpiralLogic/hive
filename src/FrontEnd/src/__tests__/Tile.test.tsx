import { TileEvents, useTileEventEmitter } from '../emitters';
import { fireEvent } from '@testing-library/preact';
import { h } from 'preact';
import { renderElement, simulateEvent } from './helpers';
import Tile from '../components/Tile';

describe('Tile', () => {

    const tileCanMove = {
        id: 1,
        playerId: 1,
        creature: 'ant',
        moves: [{ q: 1, r: 1 }],
    };
    const tileNoMove = { id: 2, playerId: 0, creature: 'fly', moves: [] };

    const createTileCanMove = () => {
        return renderElement(<Tile {...tileCanMove} />);
    };

    const createTileNoMove = () => {
        return renderElement(<Tile {...tileNoMove} />);
    };

    describe('Tile Render', () => {
        test('tile color is the player\'s color', () => {
            expect(createTileCanMove()).toHaveStyle('--color: #f64c72');
            expect(createTileNoMove()).toHaveStyle('--color: #85dcbc');
        });

        test('has creature', () => {
            expect(createTileNoMove().getElementsByTagName('use').item(0)).toHaveAttribute('href', expect.stringContaining('fly'));
            expect(createTileCanMove().getElementsByTagName('use').item(0)).toHaveAttribute('href', expect.stringContaining('ant'));
        });
    });

    test('click emits tile select event', () => {
        jest.spyOn(useTileEventEmitter(), 'emit');
        fireEvent.click(createTileCanMove());
        const expectedEvent: TileEvents = {
            type: 'start',
            tile: tileCanMove,
        };

        expect(useTileEventEmitter().emit).toHaveBeenCalledWith(expectedEvent);
    });

    describe('drag and drop', () => {
        test('Tile is draggable when there are available moves', () => {
            expect(createTileCanMove()).toHaveAttribute('draggable', 'true');
        });

        test('is *not* draggable when there are no moves available', () => {
            expect(createTileNoMove()).toHaveAttribute('draggable', 'false');
        });

        test('on drag emits start event', () => {
            jest.spyOn(useTileEventEmitter(), 'emit');
            fireEvent.dragStart(createTileCanMove());

            const expectedEvent: TileEvents = {
                type: 'start',
                tile: tileCanMove,
            };

            expect(useTileEventEmitter().emit).toHaveBeenCalledWith(expectedEvent);
        });

        test('on dragEnd emits end event', () => {
            jest.spyOn(useTileEventEmitter(), 'emit');
            fireEvent.dragEnd(createTileCanMove());
            const expectedEvent: TileEvents = {
                type: 'end',
                tile: tileCanMove,
            };

            expect(useTileEventEmitter().emit).toHaveBeenCalledWith(expectedEvent);
        });

        test('default on drop is prevented', () => {
            expect(simulateEvent(createTileCanMove(), 'drop')).toHaveBeenCalled();
            expect(simulateEvent(createTileNoMove(), 'drop')).toHaveBeenCalled();
        });
    });

    describe('Tile Snapshot', () => {
        test('can move matches current snapshot', () => {
            expect(createTileCanMove()).toMatchSnapshot();
        });

        test('no moves matches current snapshot', () => {
            expect(createTileNoMove()).toMatchSnapshot();
        });
    });
});
