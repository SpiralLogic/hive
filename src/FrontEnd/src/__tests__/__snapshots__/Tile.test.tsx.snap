// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tile Tile Snapshot can move matches current snapshot 1`] = `
<div
  class="hex tile"
  draggable="true"
  style="--color: #f64c72;"
  title="ant"
>
  <svg
    xmlns="http://www.w3.org/2000/svg"
  >
    <use
      href="/svg/creatures.svg#ant"
    />
  </svg>
</div>
`;

exports[`Tile Tile Snapshot no moves matches current snapshot 1`] = `
<div
  class="hex tile"
  draggable="false"
  style="--color: #85dcbc;"
  title="fly"
>
  <svg
    xmlns="http://www.w3.org/2000/svg"
  >
    <use
      href="/svg/creatures.svg#fly"
    />
  </svg>
</div>
`;
